src/tensor_ops.jl

```diff
@@ -263,14 +263,33 @@ Base.ctranspose(S::AllTensors) = transpose(S)
 """
 Computes the symmetric part of a second order tensor, returns a `SymmetricTensor`
 """
-@inline symmetric{dim,T}(S1::Tensor{2,dim,T}) = convert(SymmetricTensor{2,dim,T},S1)
+@generated function symmetric{order, dim, T}(t::Tensor{order, dim, T})
+    N = n_components(SymmetricTensor{order, dim})
+    rows = Int(div(sqrt(1 + 8*N), 2))
+    exps = Expr[]
+    for row in 1:rows, col in row:rows
+        if row == col
+            push!(exps, :(t.data[$(compute_index(Tensor{order, dim}, row, col))]))
+        else
+            I = compute_index(Tensor{order, dim}, row, col)
+            J = compute_index(Tensor{order, dim}, col, row)
+            push!(exps, :(0.5 * (t.data[$I] + t.data[$J])))
+        end
+    end
+    exp = Expr(:tuple, exps...)
+    return quote
+            $(Expr(:meta, :inline))
+            SymmetricTensor{order, dim, T, $N}($exp)
+        end
+end
 @inline symmetric(S1::SymmetricTensor{2}) = S1
 
+
 """
 Computes the skew-symmetric (anti-symmetric) part of a second order tensor, returns a `Tensor`
 """
 @inline skew(S1::Tensor{2}) = 0.5*(S1 - S1.')
-@inline skew{dim,T}(S1::SymmetricTensor{2,dim,T}) = zero(Tensor{2,dim,T})
+@inline skew{dim,T,M}(S1::SymmetricTensor{2,dim,T,M}) = zero(Tensor{2,dim,T,M})
```

Comment:
This `M` will not be the same for `S1` as the RHS. `M` is the number of independent components in the tensor. Why was there a need to add it?